xs/src/libslic3r/ExPolygon.cpp

```diff
@@ -195,107 +291,258 @@ ExPolygon::medial_axis(const ExPolygon &bounds, double max_width, double min_wid
 
-                //only consider the other if the next point is near us
+                //// mergeable tests
                 if (polyline.points.size() < 2 && other.points.size() < 2) continue;
                 if (!polyline.endpoints.second || !other.endpoints.second) continue;
-                if (polyline.points.back().distance_to(other.points.back()) > max_width) continue;
-                if (polyline.points.size() != other.points.size()) continue;
+                // test if the new width will not be too big if a fusion occur
+                //note that this isn't the real calcul. It's just to avoid merging lines too far apart.
+                if (
+                    ((polyline.points.back().distance_to(other.points.back())
+                        + (polyline.width.back() + other.width.back()) / 4)
+                        > max_width*1.05))
+                        continue;
+                // test if the lines are not too different in length.
+                if (abs(polyline.length() - other.length()) > max_width / 2) continue;
 
 
+                //test if we don't merge with something too different and without any relevance.
+                double coeffSizePolyI = 1;
+                if (polyline.width.back() == 0) {
+                    coeffSizePolyI = 0.1 + 0.9*get_coeff_from_angle_countour(polyline.points.back(), *this);
+                }
+                double coeffSizeOtherJ = 1;
+                if (other.width.back() == 0) {
+                    coeffSizeOtherJ = 0.1+0.9*get_coeff_from_angle_countour(other.points.back(), *this);
+                }
+                if (abs(polyline.length()*coeffSizePolyI - other.length()*coeffSizeOtherJ) > max_width / 2) continue;
+
+                //compute angle to see if it's better than previous ones (straighter = better).
                 Pointf v_poly(polyline.lines().front().vector().x, polyline.lines().front().vector().y);
                 v_poly.scale(1 / std::sqrt(v_poly.x*v_poly.x + v_poly.y*v_poly.y));
                 Pointf v_other(other.lines().front().vector().x, other.lines().front().vector().y);
                 v_other.scale(1 / std::sqrt(v_other.x*v_other.x + v_other.y*v_other.y));
                 float other_dot = v_poly.x*v_other.x + v_poly.y*v_other.y;
+
+                // Get the branch/line in wich we may merge, if possible
+                // with that, we can decide what is important, and how we can merge that.
+                // angle_poly - angle_candi =90� => one is useless
+                // both angle are equal => both are useful with same strength
+                // ex: Y => | both are useful to crete a nice line
+                // ex2: TTTTT => -----  these 90� useless lines should be discarded
+                bool find_main_branch = false;
+                int biggest_main_branch_id = 0;
+                int biggest_main_branch_length = 0;
+                for (size_t k = 0; k < pp.size(); ++k) {
```

Comment:
One way to remove these for is to have a list of intersection point and associated polyline.
Wasn't worth it for me because there's not a very high amount of pollylines as "small sections" are small.
But you're right that may be problematic time-wise in some edge cases.